### ISSUE TITLE
Injectivity lost in recursive module

### ISSUE THREAD

**Issue by stedolan** [ocaml/ocaml#10028]
This program used to typecheck in 4.11 but doesn't in 4.12 or trunk:
```ocaml
module rec A : sig
  type _ t = Foo : 'a -> 'a A.s t
  type 'a s = T of 'a
end =
  A
```
It now gives:
```
Error: In this definition, a type variable cannot be deduced
       from the type parameters.
```
(I'm not convinced that the old behaviour was right and the new one wrong, but it's definitely changed)

**Comment by gasche** [comment]
We recently fixed a hole in the injectivity computation logic due to ill-founded definitions: report in #9721, fixing PR in #9727. I suppose that the difference is due to this fix. (Which suggests that it is probably desirable, unless there are clear reasons to make the check more permissive by relaxing the fix a bit to accept a category of useful-but-safe definitions.)

**Comment by hhugo** [comment]
I'd like to point out that the error message is not very helpful, and even more in the presence of multiple type parameters

**Comment by garrigue** [comment]
In this precise case, it looks like things could be improved: since `s` is a datatype, its parameters can be marked as injective.
I will look into this.

For the error message, this is a long standing problem: at the point the error is raised, not information is available, so we get a bad message. Since there are no locations inside types, this seems hard.

**Comment by hhugo** [comment]
> For the error message, this is a long standing problem: at the point the error is raised, not information is available, so we get a bad message. Since there are no locations inside types, this seems hard

Improving the wording would already be nice. It's not clear what to look for with the current message.

**Comment by Octachron** [comment]
Fixed in 4.12 and trunk by #10029 .
